src/v/cluster/plugin_table.cc

```diff
@@ -0,0 +1,263 @@
+    }
+    return output;
+}
+
+void plugin_table::upsert_transform(transform_id id, transform_metadata meta) {
+    auto it = _name_index.find(std::string_view(meta.name()));
+    if (it != _name_index.end()) {
+        if (it->second != id) {
+            throw std::logic_error(ss::format(
+              "transform meta id={} is attempting to use a name {} which is "
+              "already registered to {}",
+              id,
+              meta.name,
+              it->second));
+        }
+    } else {
+        _name_index.emplace(meta.name, id);
+    }
+    // Topics cannot change over the lifetime of a transform, so we don't need
+    // to worry about updates specially here, additionally duplicates are
+    // allowed, the plugin_frontend asserts that there are no loops in
+    // transforms.
+    _input_topic_index[meta.input_topic].emplace(id);
+
+    for (const auto& output_topic : meta.output_topics) {
+        _output_topic_index[output_topic].emplace(id);
+    }
+    _underlying.insert_or_assign(id, std::move(meta));
+    run_callbacks(id);
+}
+
+void plugin_table::remove_transform(const transform_name& name) {
+    auto name_it = _name_index.find(std::string_view(name()));
+    if (name_it == _name_index.end()) {
+        return;
+    }
+    auto id = name_it->second;
+    auto it = _underlying.find(id);
+    vassert(
+      it != _underlying.end(),
+      "inconsistency: name index index had a record with name: {} id: {}",
+      name,
+      id);
+    const auto& meta = it->second;
+    // Delete input topic index entries
+    [this, &meta, id]() {
+        auto input_it = _input_topic_index.find(meta.input_topic);
+        vassert(
+          input_it != _input_topic_index.end()
+            && input_it->second.erase(id) > 0,
+          "inconsistency: missing transform input index entry: {} id: {}",
+          meta.input_topic,
+          id);
+        if (input_it->second.empty()) {
+            _input_topic_index.erase(input_it);
+        }
+    }();
+    for (const auto& output_topic : it->second.output_topics) {
+        auto output_it = _output_topic_index.find(output_topic);
+        vassert(
+          output_it != _output_topic_index.end()
+            && output_it->second.erase(id) > 0,
+          "inconsistency: missing transform output index entry: {} id: {}",
+          output_topic,
+          id);
+        if (output_it->second.empty()) {
+            _output_topic_index.erase(output_it);
+        }
+    }
+    _name_index.erase(name_it);
+    _underlying.erase(it);
+    run_callbacks(id);
+}
+void plugin_table::reset_transforms(plugin_table::map_t snap) {
+    name_index_t snap_name_index;
+    topic_index_t snap_input_index;
+    topic_index_t snap_output_index;
+
+    // Perform a map diff to figure out which transforms need change callbacks
+    // fired.
+    std::vector<transform_id> all_deletes;
+    std::vector<transform_id> all_changed;
+    std::vector<transform_id> all_inserted;
```

Comment:
how should we think about the size of these vectors? can they grow large?

Reply:
Generally I expect all these to be small (the number of transforms in a system) but in the same vein as making the transforms themselves btree I can make these chunked fifos too.

The only time I expect these to be large is during startup with the snapshot inserting many things.